### PR TITLE
Add proto and sidecar code for volumegroup

### PR DESCRIPTION
**Pull request by Nikhil-Ladha**
This PR adds the following codes/logic:

- Updated the csi-addons spec api
- Added volume group proto files
- Added volume group sidecar and client setup

### CODE REVIEW

**Comment by Nikhil-Ladha** [comment]
Created a separate PR for proto and sidecar code so that we can merge this and later on take the controller code once ceph code is merged.

**Comment by Nikhil-Ladha** [comment]
@nixpanic @Madhu-1 @Rakshith-R can you please take a look.

File: internal/proto/volumegroup.proto [hunk 1]
```diff
@@ -0,0 +1,111 @@
+syntax = "proto3";
+package proto;
+
+option go_package = "github.com/csi-addons/kubernetes-csi-addons/internal/proto";
```
Comment (Rakshith-R) [inline internal/proto/volumegroup.proto:4]: Can you check again ? 
I don't think this matches csi-addons spec or cephcsi implementation
https://github.com/csi-addons/spec/tree/main/volumegroup#csi-addons-operation-volumegroup
https://github.com/ceph/ceph-csi/blob/devel/internal/csi-addons/rbd/volumegroup.go

Let's implement everything here too and ignore information that's not needed right now at controller level.

File: internal/proto/volumegroup.proto [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+  rpc CreateVolumeGroup(CreateVolumeGroupRequest)
+      returns (CreateVolumeGroupResponse) {}
+  // ModifyVolumeGroupMembership RPC call to modify a volume group.
+  rpc ModifyVolumeGroupMembership(ModifyVolumeGroupMembershipRequest)
+      returns (ModifyVolumeGroupMembershipResponse) {}
+  // DeleteVolumeGroup RPC call to delete a volume group.
+  rpc DeleteVolumeGroup(DeleteVolumeGroupRequest)
+      returns (DeleteVolumeGroupResponse) {}
+  // CreateVolumeGroup RPC call to get a volume group.
+  rpc ControllerGetVolumeGroup(ControllerGetVolumeGroupRequest)
```
Comment (Madhu-1) [inline internal/proto/volumegroup.proto:19]: update the comment on this RPC

File: internal/proto/volumegroup.proto [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+  // Secrets required by the driver to complete the
+  // get request.
+  string secret_name = 2;
+  string secret_namespace = 3;
+}
+// ControllerGetVolumeGroupResponse is the response to
+// send when volumeGroup information was successfully gathered.
+message ControllerGetVolumeGroupResponse {
+}
```
Comment (Madhu-1) [inline internal/proto/volumegroup.proto:111]: GitHub is complaining that no newline at the end of the file

File: internal/sidecar/service/volumegroup.go [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+	vgReq := &csiVolumeGroup.ControllerGetVolumeGroupRequest{
+		VolumeGroupId: req.GetVolumeGroupId(),
+		Secrets:       data,
+	}
+
+	_, err = vg.controllerClient.ControllerGetVolumeGroup(ctx, vgReq)
+	if err != nil {
+		klog.Errorf("Failed to get volume group: %v", err)
+		return nil, err
+	}
```
Comment (Madhu-1) [inline internal/sidecar/service/volumegroup.go:156]: what is the use of this one if we don't fetch the information and surface it to the caller?

File: internal/proto/volumegroup.proto (hunk 1 above)
Comment (Nikhil-Ladha) [reply internal/proto/volumegroup.proto:4]: Updated it now to include everything as per the sepc defined in the csi-addons. Please check now.

**Review by Madhu-1 — DISMISSED** [review]

File: internal/proto/volumegroup.proto [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+// CreateVolumeGroupResponse holds the information to send when
+// volumeGroup is successfully created.
+message CreateVolumeGroupResponse {
+  // Contains the identifier of the newly created volume group.
+  // This field is REQUIRED.
+  VolumeGroup volume_group = 1;
+}
+// Information about a specific volumeGroup.
+message VolumeGroup {
```
Comment (nixpanic) [inline internal/proto/volumegroup.proto:53]: Is there a reason to not use the CSI-Addons VolumeGroup type? Duplicating it looks unneeded to me.

File: internal/proto/volumegroup.proto [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+}
+// ControllerGetVolumeGroupResponse is the response to
+// send when volumeGroup information was successfully gathered.
+message ControllerGetVolumeGroupResponse {
+  // This field is REQUIRED.
+  VolumeGroup volume_group = 1;
+}
+// ListVolumeGroupsRequest holds the required
+// information to get information on list of volume groups.
+message ListVolumeGroupsRequest {
```
Comment (nixpanic) [inline internal/proto/volumegroup.proto:142]: Are you planning to use this function? No need to include functions that you won't use anyway.

File: internal/proto/volumegroup.proto (hunk 5 above)
Comment (Nikhil-Ladha) [reply internal/proto/volumegroup.proto:53]: We would have to tweak the Volumes type to use in CSI-Addons, we can't use the CSI volume type. That's the reaspn I updated it.

File: internal/proto/volumegroup.proto (hunk 6 above)
Comment (Nikhil-Ladha) [reply internal/proto/volumegroup.proto:142]: I am not but as per the above comments, I included it.
I will remove it in my controller PR, or shall I remove it here itself?

Comment (nixpanic) [reply internal/proto/volumegroup.proto:142]: I do not see a point in including anything that won't be used, that can only cause confusion. Do not add things that will be removed in a follow-up commit/PR.

The files under `internal/proto/` are only for the communication between the controller and sidecar, it should be as simple and minimal as possible.

Comment (Nikhil-Ladha) [reply internal/proto/volumegroup.proto]: Updated the code to remove the ListVolumeGroups RPC.

File: internal/proto/volumegroup.proto (hunk 5 above)
Comment (Nikhil-Ladha) [reply internal/proto/volumegroup.proto:50]: Done

**Review by Rakshith-R — APPROVED** [review]
LGTM

**Comment by nixpanic** [comment]
Looks good to me. @Madhu-1 do you want to have a final look?

**Review by Madhu-1 — APPROVED** [review]

**Comment by Nikhil-Ladha** [comment]
@mergifyio refresh

**Comment by Madhu-1** [comment]
@mergifyio rebase

**Pull request merged**